contracts/src/token/erc721/extensions/wrapper.rs

```diff
@@ -0,0 +1,887 @@
+
+        let initial_contract_balance = erc721_contract
+            .sender(alice)
+            .balance_of(contract.address())
+            .motsu_unwrap();
+
+        assert!(contract
+            .sender(alice)
+            .withdraw_to(alice, token_ids.clone())
+            .motsu_expect("should withdraw"));
+
+        for token_id in token_ids {
+            erc721_contract.assert_emitted(&erc721::Transfer {
+                from: contract.address(),
+                to: alice,
+                token_id,
+            });
+
+            contract.assert_emitted(&erc721::Transfer {
+                from: alice,
+                to: Address::ZERO,
+                token_id,
+            });
+        }
+
+        assert_eq!(
+            erc721_contract.sender(alice).balance_of(alice).unwrap(),
+            initial_balance + U256::from(tokens)
+        );
+
+        assert_eq!(
+            contract.sender(alice).erc721.balance_of(alice).unwrap(),
+            initial_wrapped_balance - U256::from(tokens)
+        );
+
+        assert_eq!(
+            erc721_contract
+                .sender(contract.address())
+                .balance_of(contract.address())
+                .motsu_unwrap(),
+            initial_contract_balance - U256::from(tokens)
+        );
+    }
+
+    #[motsu::test]
+    fn on_erc721_received_reverts_when_invalid_operator(
+        contract: Contract<Erc721WrapperTestExample>,
+        erc721_contract: Contract<Erc721>,
+        alice: Address,
+    ) {
+        let token_id = random_token_ids(1)[0];
+
+        contract.init(alice, |contract| {
+            contract.wrapper.underlying.set(erc721_contract.address());
+        });
+
+        let invalid_operator = alice;
+
+        let err = contract
+            .sender(invalid_operator)
+            .on_erc721_received(
+                invalid_operator,
+                alice,
+                token_id,
+                vec![].into(),
+            )
+            .motsu_expect_err("should return Error::UnsupportedToken");
+
+        assert!(matches!(
+            err,
+            Error::UnsupportedToken(ERC721UnsupportedToken { token })
+                if token == invalid_operator
+        ));
+    }
+
+    #[motsu::test]
+    fn on_erc721_received_works(
+        contract: Contract<Erc721WrapperTestExample>,
+        erc721_contract: Contract<Erc721>,
+        alice: Address,
+    ) {
+        let token_id = random_token_ids(1)[0];
+
+        contract.init(alice, |contract| {
+            contract.wrapper.underlying.set(erc721_contract.address());
+        });
+
+        let initial_wrapped_balance =
+            contract.sender(alice).erc721.balance_of(alice).motsu_unwrap();
+
+        let operator = erc721_contract.address();
+        let interface_id = contract
+            .sender(operator)
+            .on_erc721_received(operator, alice, token_id, vec![].into())
+            .motsu_expect("should handle ERC721Received");
```

Comment:
```suggestion
        let operator = alice;
        let interface_id = contract
            .sender(erc721_contract.address())
            .on_erc721_received(operator, alice, token_id, vec![].into())
            .motsu_expect("should handle ERC721Received");
```
note: `operator` represents the original initiator of the tx, i.e. the one that called some function on `erc721_contract` which caused it to invoke the wrapper's `on_erc721_received`. 
A more meaningful `operator` would be `alice`.
Pointing this out to avoid confusion

Reply:
Following the [Solidity implementation](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v5.2.0/contracts/token/ERC721/extensions/ERC721Wrapper.sol#L76) it should be the underlying token... Maybe there is a bug in this implementation because for me it does not have sense.

Reply:
Yeah yeah, you got it correct - the underlying token is the `msg::sender`, as is in this test case, and the `operator` is the account that initiated the original transaction on the underlying token that caused it to invoke `on_erc721_received`.
To give an example:
- Alice (operator) invokes some operation `X` on `erc721_contract`
- during the `X` operation, `erc721_contract` safe-transfers some token to `wrapper`
    - because this is a "safe" transfer, `erc721_contract` invokes `wrapper.on_erc721_received`
        - operator is now Alice, because she's the one who invoked `X`, while the `msg::sender` is `erc721_contract`.

Reply:
Oh yeah, you are right, it was too late for me 😅